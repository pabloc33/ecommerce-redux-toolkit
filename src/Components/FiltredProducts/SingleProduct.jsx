import { useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import { useParams } from "react-router-dom";
import { Tooltip, Button } from "@material-tailwind/react";
import { addToCart } from "../../features/slices/cartSlice";

const SingleProduct = () => {
  const product = useSelector((state) => state.products.singleProduct);
  const productSize = product[0].size ? product[0].size[0] : "";
  const productColor = product[0].color[0];
  const [size, setSize] = useState(productSize);
  const [color, setColor] = useState(productColor);

  const { id } = useParams();

  const dispatch = useDispatch();

  return (
    <div>
      {product
        .filter((product) => product.id === id)
        .map((item, index) => {
          return (
            <div
              key={index}
              className="flex flex-col md:flex-row items-center md:py-10"
            >
              <div className="md:pl-6 lg:pl-12 xl:pl-20 grow-[2]">
                <img
                  className="w-full md:max-w-[450px] lg:max-w-[550px] xl:max-w-[650px] h-auto rounded-lg"
                  //src={toString(item.img)}
                  src={item.img}
                  alt={item.name}
                />
              </div>

              <div className="grow-[3] mt-6 md:mt-0 md:pl-6 lg:pl-12 xl:pl-20">
                <div>
                  <div className="max-w-lg">
                    <h5 className="text-2xl md:text-3xl lg:text-4xl font-inter font-bold tracking-normal leading-normal">
                      {item.name}
                    </h5>

                    <p className="text-orange-700 text-xl md:text-2xl lg:text-3xl font-inter font-bold tracking-normal leading-relaxed pb-4">
                      15% OFF
                    </p>
                    <p className="text-gray-600 text-xl md:text-2xl lg:text-3xl font-inter font-bold tracking-normal leading-relaxed pb-4">
                      {item.text}
                    </p>

                    <div className="pb-4">
                      {item.size ? (
                        <div>
                          <label
                            htmlFor="size"
                            className="block mb-2 text-sm md:text-base font-medium text-gray-900 dark:text-white"
                          >
                            Pick a size
                          </label>
                          <select
                            name="size"
                            id="size"
                            value={size}
                            onChange={(e) => setSize(e.target.value)}
                            className="w-full md:max-w-md lg:max-w-lg bg-gray-50 border border-gray-300 text-gray-900 text-sm md:text-base rounded-lg focus:ring-blue-500 focus:border-blue-500 p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
                          >
                            {item.size.map((item, index) => {
                              return (
                                <option key={index} value={item}>
                                  {item}
                                </option>
                              );
                            })}
                          </select>
                        </div>
                      ) : (
                        <div>
                          <label
                            htmlFor="size"
                            className="block mb-2 text-sm md:text-base font-medium text-gray-900 dark:text-white"
                          >
                            Pick a size
                          </label>
                          <select
                            name="size"
                            id="size"
                            value={size}
                            onChange={(e) => setSize(e.target.value)}
                            className="w-full md:max-w-md lg:max-w-lg bg-gray-50 border border-gray-300 text-gray-900 text-sm md:text-base rounded-lg focus:ring-blue-500 focus:border-blue-500 p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
                          >
                            {item?.size?.map((item, index) => {
                              return (
                                <option key={index} value={item}>
                                  {item}
                                </option>
                              );
                            })}
                          </select>
                        </div>
                      )}
                    </div>

                    <div className="pb-4">
                      <label
                        htmlFor="color"
                        className="block mb-2 text-sm md:text-base font-medium text-gray-900 dark:text-white"
                      >
                        Pick a color
                      </label>
                      <select
                        id="color"
                        name="color"
                        value={color}
                        onChange={(e) => setColor(e.target.value)}
                        className="w-full md:max-w-md lg:max-w-lg bg-gray-50 border border-gray-300 text-gray-900 text-sm md:text-base rounded-lg focus:ring-blue-500 focus:border-blue-500 p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
                      >
                        {item.color.map((color, index) => {
                          return (
                            <option key={index} value={color}>
                              {color}
                            </option>
                          );
                        })}
                      </select>
                    </div>

                    <Tooltip content="Add to Cart" placement="bottom">
                      <Button
                        aria-label="addToCart-btn"
                        color="gray"
                        size="lg"
                        variant="outlined"
                        ripple={true}
                        onClick={() =>
                          dispatch(
                            addToCart({
                              id: item.id,
                              name: item.name,
                              img: item.img,
                              text: item.text,
                              size: size,
                              color: color,
                              price: item.price,
                              amount: 1,
                              totalPrice: item.price,
                            })
                          )
                        }
                      >
                        Add to Cart
                      </Button>
                    </Tooltip>
                  </div>
                </div>
              </div>
            </div>
          );
        })}
    </div>
  );
};

export default SingleProduct;
