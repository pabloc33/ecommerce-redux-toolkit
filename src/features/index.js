export * from "./slices/sliderSlice";
export * from "./slices/productsSlice";
export * from "./slices/cartSlice";
export * from "./slices/authSlice";
