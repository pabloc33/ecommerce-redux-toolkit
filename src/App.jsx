import { BrowserRouter, Routes, Route } from "react-router-dom";
import Main from "./Components/Main/Main";
import FiltredProducts from "./Components/FiltredProducts/FiltredProducts";
import SingleProduct from "./Components/FiltredProducts/SingleProduct";
import Login from "./Components/Login/Login";
import { useSelector } from "react-redux";
import "./App.css";

function App() {
  const user = useSelector((state) => state.user.user);
  const { authUser } = user;

  return (
    <>
      <div className="">
        <BrowserRouter>
          <Routes>
            <Route path="/" element={authUser ? <Main /> : <Login />} />
            <Route path="/login" element={<Login />} />
            <Route
              path="/filteredProducts/:type"
              element={<FiltredProducts />}
            />
            <Route
              path="/filteredProducts/:type/:id"
              element={<SingleProduct />}
            />
          </Routes>
        </BrowserRouter>
      </div>
    </>
  );
}

export default App;
