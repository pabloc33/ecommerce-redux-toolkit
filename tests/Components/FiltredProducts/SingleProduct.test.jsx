import { fireEvent, render, screen } from "@testing-library/react";
import SingleProduct from "../../../src/Components/FiltredProducts/SingleProduct";
import { configureStore } from "@reduxjs/toolkit";
import { Provider, useDispatch } from "react-redux";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import { productSlice } from "../../../src/features";
import { expectedState } from "../../fixtures/productsFixture";

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: jest.fn(),
}));

const store = configureStore({
  reducer: {
    products: productSlice.reducer,
  },
  preloadedState: {
    products: expectedState,
  },
});

describe("Pruebas en <SingleProduct />", () => {
  test("debe de mostrarse correactamente con valores por defecto", () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/filteredProducts/T-Shirts/1"]}>
          <Routes>
            <Route
              path="/filteredProducts/:type/:id"
              element={<SingleProduct />}
            />
          </Routes>
        </MemoryRouter>
      </Provider>
    );

    //screen.debug();
    expect(container).toMatchSnapshot();
  });

  test("debe de seleccionar Add to Cart y llamar addToCart", () => {
    const dispatch = jest.fn();
    useDispatch.mockReturnValue(dispatch);

    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/filteredProducts/T-Shirts/1"]}>
          <Routes>
            <Route
              path="/filteredProducts/:type/:id"
              element={<SingleProduct />}
            />
          </Routes>
        </MemoryRouter>
      </Provider>
    );

    //screen.debug();
    const addToCartBtn = screen.getByLabelText("addToCart-btn");
    fireEvent.click(addToCartBtn);

    expect(dispatch).toHaveBeenCalledWith({
      type: "cart/addToCart",
      payload: {
        amount: 1,
        color: "black",
        id: "1",
        img: {},
        name: "Casual T-Shirt 1",
        price: 45,
        size: "XL",
        text: "Fashion never stops. There is always the new project, the new opportunity. The important thing is to take your time and not get stressed. I just want to do what I do.",
        totalPrice: 45,
      },
    });
  });
});
