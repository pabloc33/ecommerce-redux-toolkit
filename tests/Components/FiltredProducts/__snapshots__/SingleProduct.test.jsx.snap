// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SingleProduct /> debe de mostrarse correactamente con valores por defecto 1`] = `
<div>
  <div>
    <div
      class="flex flex-col md:flex-row items-center md:py-10"
    >
      <div
        class="md:pl-6 lg:pl-12 xl:pl-20 grow-[2]"
      >
        <img
          alt="Casual T-Shirt 1"
          class="w-full md:max-w-[450px] lg:max-w-[550px] xl:max-w-[650px] h-auto rounded-lg"
          src="[object Undefined]"
        />
      </div>
      <div
        class="grow-[3] mt-6 md:mt-0 md:pl-6 lg:pl-12 xl:pl-20"
      >
        <div>
          <div
            class="max-w-lg"
          >
            <h5
              class="text-2xl md:text-3xl lg:text-4xl font-inter font-bold tracking-normal leading-normal"
            >
              Casual T-Shirt 1
            </h5>
            <p
              class="text-orange-700 text-xl md:text-2xl lg:text-3xl font-inter font-bold tracking-normal leading-relaxed pb-4"
            >
              15% OFF
            </p>
            <p
              class="text-gray-600 text-xl md:text-2xl lg:text-3xl font-inter font-bold tracking-normal leading-relaxed pb-4"
            >
              Fashion never stops. There is always the new project, the new opportunity. The important thing is to take your time and not get stressed. I just want to do what I do.
            </p>
            <div
              class="pb-4"
            >
              <div>
                <label
                  class="block mb-2 text-sm md:text-base font-medium text-gray-900 dark:text-white"
                  for="size"
                >
                  Pick a size
                </label>
                <select
                  class="w-full md:max-w-md lg:max-w-lg bg-gray-50 border border-gray-300 text-gray-900 text-sm md:text-base rounded-lg focus:ring-blue-500 focus:border-blue-500 p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
                  id="size"
                  name="size"
                >
                  <option
                    value="XL"
                  >
                    XL
                  </option>
                  <option
                    value="L"
                  >
                    L
                  </option>
                  <option
                    value="M"
                  >
                    M
                  </option>
                  <option
                    value="S"
                  >
                    S
                  </option>
                </select>
              </div>
            </div>
            <div
              class="pb-4"
            >
              <label
                class="block mb-2 text-sm md:text-base font-medium text-gray-900 dark:text-white"
                for="color"
              >
                Pick a color
              </label>
              <select
                class="w-full md:max-w-md lg:max-w-lg bg-gray-50 border border-gray-300 text-gray-900 text-sm md:text-base rounded-lg focus:ring-blue-500 focus:border-blue-500 p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
                id="color"
                name="color"
              >
                <option
                  value="black"
                >
                  black
                </option>
                <option
                  value="gray"
                >
                  gray
                </option>
                <option
                  value="green"
                >
                  green
                </option>
              </select>
            </div>
            <button
              aria-label="addToCart-btn"
              class="align-middle select-none font-sans font-bold text-center uppercase transition-all disabled:opacity-50 disabled:shadow-none disabled:pointer-events-none text-sm py-3.5 px-7 rounded-lg border border-gray-900 text-gray-900 hover:opacity-75 focus:ring focus:ring-gray-300 active:opacity-[0.85] gap-3"
              type="button"
            >
              Add to Cart
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
